### ISSUE TITLE
Processing a large compile_commands.json takes too long

### ISSUE THREAD

**Issue by akbyrd** [microsoft/vscode-cpptools#12825]
### Environment

- OS and Version: Win10 22H2 19045
- VS Code Version: 1.94 
- C/C++ Extension Version: 1.21.6
- If using SSH remote, specify OS of remote machine: N/A

### Bug Summary and Steps to Reproduce

Bug Summary: The extension gets stuck initializing Intellisense more or less permanently when loading a compilation database with a large number of file entries. The extension is unusable while this occurs. It's a single threaded process that appears to process every file in the compilation database before enabling any LSP features for currently opened files.

![Image](https://github.com/user-attachments/assets/68e62d80-cad2-4855-9b1a-fe244158bbbc)

Steps to reproduce:
1. Clone UnrealEngine (use a shallow clone to make it faster)
    a. Sorry, I know this is a pain for a repro. There's a small, self-contained partial repro I can't share publicly if there's a private channel available.
1. Open the folder with vscode
1. Run `GenerateProjectFiles.bat -VSCode`
1. Close and re-open vscode (still as a folder, you don't need to use the workspace)
1. Open a file (e.g. `AssertionMacroTests.cpp`)
1. Notice Intellisense is initializing
1. Wait a while
1. Notice Intellisense is still initializing
1. Attempt to "go to definition" on a `#include` statement
1. Notice no code navigation or code completion operations work

Expected behavior:
* Intellisense and LSP operations for the currently active file are prioritized above all other files.
* Beginning normal LSP operations is not blocked on doing any operation on every single file in the directory.

When not using compile commands the tag parser fallback behaves correctly. It will immediately parse the current open file. Similarly, clangd handles the current file before doing background indexing.

It's also possible to observe the bad behavior with just the .vscode directory and not the entire engine. When the actual code files are missing the `C/C++ Configuration Warnings` output will be spammed with errors like this:
```
[10/9/2024, 9:37:04 AM] Unable to locate forced include: D:\UnrealEngine\Engine\Intermediate\Build\Win64\x64\UnrealEditorGPF\Development\UnrealEd\SharedPCH.UnrealEd.Cpp20.h
```
While this is occurring, Intellisense is in the "initializing" state indicating that it is processing every single file in the database before the init completes and LSP operations in the current file begin to function.

### Configuration and Logs

```
{
    "configurations": [
        {
            "name": "Win32",
            "compilerPath": "C:\\Program Files\\Microsoft Visual Studio\\2022\\Professional\\VC\\Tools\\MSVC\\14.38.33130\\bin\\Hostx64\\x64\\cl.exe",
            "cStandard": "c17",
            "cppStandard": "c++20",
            "intelliSenseMode": "msvc-x64",
            "compileCommands": "D:\\UnrealEngine\\.vscode\\compileCommands_Default.json"
        }
    ],
    "version": 4
}
```

I don't think I can get `C/C++: Log Diagnostics`. The command doesn't do anything. Looks like it's an LSP request? And since the LSP is essentially stalled it's not returning a response.

```
loggingLevel: Debug
LSP: (received) cpptools/initialize (id: 1)
LSP: (invoked) cpptools/initialize (id: 1)
cpptools version (TypeScript): 1.21.6
cpptools version (native): 1.21.6.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 1)
LSP: (received) cpptools/queryCompilerDefaults (id: 2)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 2)
LSP: Sending response (id: 2)
LSP: (received) cpptools/didChangeCppProperties (id: 3)
LSP: (invoked) cpptools/didChangeCppProperties (id: 3)
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received - deferred) textDocument/didOpen: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
LSP: Message ignored due to no registered handler: $/setTrace
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 4)
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 5)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 6)
LSP: (received) cpptools/didChangeSettings
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 7)
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 8)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 4)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 5)
LSP: $/cancelRequest (cpptools/getFoldingRanges, id: 6)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 7)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 9)
Code browsing service initialized
Attempting to get defaults from C++ compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
Attempting to get defaults from C compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
```

### Other Extensions

I've repro'd on Insiders with no other extensions.

**Comment by sean-mcmanus** [comment]
@akbyrd Did you repro this with 1.22.8 too? That has some fixes for some potential deadlocks.

It appears to be stuck from `(invoked) cpptools/didChangeCppProperties`

**Comment by akbyrd** [comment]
Yea, looks like the insiders version I have is 1.22.8 and I was able to repro there. The logging is a little different though. I see a continuous stream of
```
Attempting to get defaults from C++ compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
```

It doesn't seem stalled, it just seems like it's doing a boatload of work before enabling LSP operations.

```
loggingLevel: Debug
LSP: (received) cpptools/initialize (id: 1)
LSP: (invoked) cpptools/initialize (id: 1)
cpptools version (TypeScript): 1.22.8
cpptools version (native): 1.22.8.0
Autocomplete is enabled.
Error squiggles are enabled if all header dependencies are resolved.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 1)
LSP: (received) cpptools/queryCompilerDefaults (id: 2)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 2)
Compiler info database not connected - skipping load.
Compiler info database not connected - skipping load.
LSP: Sending response (id: 2)
LSP: (received) cpptools/didChangeCppProperties (id: 3)
LSP: (invoked) cpptools/didChangeCppProperties (id: 3)
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received) cpptools/didChangeSettings
LSP: (received - deferred) textDocument/didOpen: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 4)
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 5)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 6)
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 7)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 4)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 5)
LSP: $/cancelRequest (cpptools/getFoldingRanges, id: 6)
LSP: Message ignored due to no registered handler: $/setTrace
LSP: (received) cpptools/didChangeSettings
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 8)
LSP: $/cancelRequest (cpptools/getCodeActions, id: 7)
LSP: (received - deferred) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 9)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 10)
Code browsing service initialized
Populating file name cache...
Done populating filename cache. Elapsed time: 0 ms
Attempting to get defaults from C compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
Attempting to get defaults from C++ compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
Attempting to get defaults from C++ compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
Attempting to get defaults from C++ compiler in "compilerPath" property: 'C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\bin\Hostx64\x64\cl.exe'
...
```

**Comment by sean-mcmanus** [comment]
@akbyrd Yeah, I think it's processing the compile_commands command. This was improved with https://github.com/microsoft/vscode-cpptools/issues/10029. How many compile commands entries are there? How long are you seeing it take to finish processing that?

**Comment by akbyrd** [comment]
77,727 in one workspace with just the engine
108,140 in another with the projects I need to work in

Big numbers, and I'll be the first to complain about it, but it is what it is. I can't really pare it down since I'm poking at foundational bits that everything uses.

I think the important requirement here is that Intellisense and LSP operations for the current file can't depend on work being performed in _all_ files, only its actual dependencies. That might still be a large number of files, and that's ok, as long as it's not artificially gated by _everything_.

I don't think this was always an issue. I've been working in this codebase for a few years and I think the extension would fallback to the tag parser if Intellisense was still trying to catch up. I vaguely recall that I'd have basic operations that weren't entirely accurate for a while in a new workspace, then it would catch up after a while. Currently I get no operations at all, everything stalls with a loading bar until the init phase is complete.

> How long are you seeing it take to finish processing that?
I haven't seen it finish yet. I'll let it go overnight tonight to see if it completes. How can I see the progress? Is there a log file with progress and timestamps so I can estimate the completion time if it doesn't finish?

**Comment by akbyrd** [comment]
After letting it sit for a while it did get past initializing. Without timestamps I can't say how long it took, but more than 1 hour and less than 2.75 hours.

I set the logging to "7" instead of "Debug" which I saw you mention elsewhere. I see about 100,000
```
LSP: Sending server -> client notification: cpptools/updateTrustedCompilersList
```

then a few
```
Folder: D:\BLAH\ will be indexed
```

then all the pending LSP operations finally begin to process
```
LSP: (queued) textDocument/didOpen: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
LSP: (queued) cpptools/didChangeVisibleTextEditors
LSP: (queued) cpptools/didChangeVisibleTextEditors
LSP: (queued) cpptools/didChangeVisibleTextEditors
LSP: (queued) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 4)
LSP: (queued) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 5)
LSP: (queued) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 6)
LSP: (queued) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 7)
LSP: (queued) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 8)
LSP: (queued) cpptools/getFoldingRanges: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 9)
LSP: (queued) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 10)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 7)
LSP: Sending response (id: 3)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 10)
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: Sending server -> client notification: cpptools/reportTagParseStatus
LSP: Sending server -> client notification: cpptools/reportStatus
Discovering files...
LSP: (invoked) cpptools/didChangeSettings
LSP: Sending server -> client notification: cpptools/reportStatus
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
Enhanced Colorization is enabled.
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: (invoked) cpptools/didChangeSettings
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
Enhanced Colorization is enabled.
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: (invoked) cpptools/didChangeSettings
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
tag parsing file: D:\UnrealEngine\Engine\Source\Runtime\Core\Tests\Misc\AssertionMacrosTest.cpp
Enhanced Colorization is enabled.
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: (invoked) cpptools/didChangeSettings
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
Enhanced Colorization is enabled.
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: (invoked) cpptools/queryCompilerDefaults (id: 11)
LSP: Sending response (id: 11)
LSP: (invoked) textDocument/didOpen: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
  Processing folder (non-recursive): C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\PROFESSIONAL\VC\TOOLS\MSVC\14.29.30133\INCLUDE
process_path: C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\PROFESSIONAL\VC\TOOLS\MSVC\14.29.30133\INCLUDE
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: Sending server -> client notification: cpptools/IntelliSenseSetup
CodeSearch (prefix phase): Time(2ms) ResultsCount(6) Indexed(t) Fuzzy(f) SingleFile (t) Scoped(f) ExcludedImplicit(f)
CodeSearch: TotalResultsCount(6) TotalTime(2ms)
  Processing folder (recursive): C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.26100.0\CPPWINRT\
process_path: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.26100.0\CPPWINRT\
CodeSearch: LoadResultsTime(14ms)
CodeSearch: QuerySymbolsTotalTime(17ms)
LSP: Sending response (id: 7)
CodeSearch (prefix phase): Time(1ms) ResultsCount(6) Indexed(t) Fuzzy(f) SingleFile (t) Scoped(f) ExcludedImplicit(f)
CodeSearch: TotalResultsCount(6) TotalTime(1ms)
CodeSearch: LoadResultsTime(7ms)
CodeSearch: QuerySymbolsTotalTime(9ms)
LSP: Sending response (id: 10)
tag parsing file: D:\UnrealEngine\Engine\Source\Runtime\Core\Tests\Misc\AssertionMacrosTest.cpp
Intellisense update pending for: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
LSP: Sending server -> client notification: cpptools/reportStatus
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 4)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 5)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (id: 6)
```

then some more processing
```
Populating file name cache...
  Processing folder (recursive): C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.26100.0\UCRT\
```

```
Done populating filename cache. Elapsed time: 812 ms
    Memory usage delta: 142 MB
    Number of files: 6589
    Number of parent/child relationships: 5854
sending compilation args for D:\UnrealEngine\Engine\Source\Runtime\Core\Tests\Misc\AssertionMacrosTest.cpp
  include: D:\UNREALENGINE\ENGINE\SOURCE
...
  system include: C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\PROFESSIONAL\VC\TOOLS\MSVC\14.38.33130\INCLUDE
...
  preinclude: D:\UNREALENGINE\ENGINE\INTERMEDIATE\BUILD\WIN64\X64\ACEXEPARSERGPF\DEVELOPMENT\CORE\DEFINITIONS.H
  stdver: ms_c++20
  intelliSenseMode: windows-msvc-x64
IntelliSense TU acquired for: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp
Starting IntelliSense update for: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp (needs_reparse==true)
Processing partial IntelliSense update for: file:///d%3A/UnrealEngine/Engine/Source/Runtime/Core/Tests/Misc/AssertionMacrosTest.cpp [0-850]
LSP: Sending server -> client notification: cpptools/reportTagParseStatus
```
```
  Discovering files: 313511 file(s) processed
  0 file(s) removed from database
LSP: Sending server -> client notification: cpptools/logTelemetry
Done discovering files.
LSP: Sending server -> client notification: cpptools/reportStatus
Parsing open files...
LSP: Sending server -> client notification: cpptools/reportStatus
Populating include completion cache.
Parsing remaining files...
tag parsing file: D:\UnrealEngine\Engine\Source\Runtime\Core\Public\Misc\AssertionMacros.h
...
```

**Comment by sean-mcmanus** [comment]
@akbyrd Was the 100k updateTrustedCompilersList with 1.22.8? That was supposed to be fixed by https://github.com/microsoft/vscode-cpptools/issues/12078 .

I think we only tested performance for a 10k compile_commands and 70-110k might require additional optimization or algorithm changes.

**Comment by akbyrd** [comment]
No, it was my retail install that I let run (1.21.6). Is there any other helpful data I can gather?

**Comment by sean-mcmanus** [comment]
@akbyrd If you can try with 1.22.8 (which we're planning to release soon as 1.22.9 non-pre-release) and let us know if (how much) it is faster that would help. But otherwise, I think we have enough info.

**Comment by akbyrd** [comment]
I let 1.22.8 grind away on the 108k workspace and it took around 2.75 hours (+/- 30 minutes)

**Comment by sean-mcmanus** [comment]
@akbyrd Okay, sounds like "no improvement".

**Comment by sean-mcmanus** [comment]
@akbyrd Also, FYI, the Visual Studio 2022 product has specific optimizations and functionality for Unreal Engine that don't exist in VS Code, so you may want to consider that instead.

**Comment by akbyrd** [comment]
Thanks for the tip. I'm using that as a fallback for the moment. But the goal is to stick when vscode as vs is much too slow for daily use.

I noticed it does some initial parsing for headers for the first minute or so and then Intellisense operations work so it doesn't seem to get blocked up on the initial. But the build tool for Unreal produces some sort of Intellisense data in advance, but I haven't looked into exactly what that data is yet. So it might be pre-populating some of the data that vscode has to figure out for itself.

**Comment by sharadhr** [comment]
Hi there! 

I noticed there are improvements in #10029 which is nice to see. 

That being said, the C++ IntelliSense engine completely crashes on databases that are ~1.5 orders of magnitude larger. For instance, I work in Android upstream, which generates a `compile_commands.json` file of ~3.5 GiB. I can't share an actual dump since this is internal, but this can be repro with the following: 

1. Create an [AOSP checkout](https://source.android.com/docs/setup/start) (you'll need a _huge_ amount of free disk space, ~1 TiB). 

1. Once the checkout is fully set up, `cd` to it, and execute
   ```
   source build/envsetup.sh
   lunch aosp_cf_x86_64_only_phone-aosp_current-userdebug
   m SOONG_GEN_COMPDB=1 SOONG_LINK_COMPDB_TO=$ANDROID_BUILD_TOP nothing
   ```
   which will create a `compile_commands.json` file at the top level of the checkout. This is a symlink to `$ANDROID_BUILD_TOP/out/soong/development/ide/compdb/compile_commands.json` (`$ANDROID_BUILD_TOP` is set in the current shell environment after the `source` and `lunch` commands above), 

1. Ensure that `"C_Cpp.default.compileCommands": "${workspaceFolder}/compile_commands.json"` is set in the workspace/user settings, and open VS Code.

1. Observe the IntelliSense engine just crash repeatedly, and give up after ~5 attempts. 

I have had to switch to `clangd`, which brings its own set of problems. I'd love to see this fixed.

**Comment by sean-mcmanus** [comment]
@sharadhr What do you mean by "IntelliSense engine just crashed"? Which error message are you seeing? We have the main cpptools process and separate cpptools-srv IntelliSense processes per-TU. Are you able to provide a call stack of the crash or does one of those processes appear to crash from running out of memory? What OS are you on?

**Comment by sharadhr** [comment]
@sean-mcmanus thanks for responding! I'm seeing the following toast message:

<img width="802" height="384" alt="Image" src="https://github.com/user-attachments/assets/af27f36f-73b6-4355-9535-e17915684259" />

And the following logs in the `C/C++` log window (any internal paths are redacted):

```
loggingLevel: 6
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received) cpptools/preinitialize (id: 1)
LSP: (invoked) cpptools/preinitialize (id: 1)
LSP: Sending response (id: 1)
LSP: (received) cpptools/initialize (id: 2)
LSP: (invoked) cpptools/initialize (id: 2)
cpptools version (TypeScript): 1.32.2
cpptools version (native): 1.32.2.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 2)
LSP: (received) cpptools/queryCompilerDefaults (id: 3)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 3)
Querying compiler for default C++ language standard using command line: /usr/bin/clang -x c++ -E -dM /dev/null
Detected language standard version: c++17
Querying compiler's default target using command line: "/usr/bin/clang" -dumpmachine
Compiler returned default target value: x86_64-pc-linux-gnu
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c17 -m64 -Wp,-v -fno-blocks -E -dM -x c /dev/null
Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/clang'
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c++17 -m64 -Wp,-v -fno-blocks -E -dM -x c++ /dev/null
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/clang'
LSP: Sending response (id: 3)
Processed c_cpp_properties.json in 0.007664449s
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 5)
LSP: (invoked) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) cpptools/getCodeActions: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 6)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 7)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received) cpptools/preinitialize (id: 1)
LSP: (invoked) cpptools/preinitialize (id: 1)
LSP: Sending response (id: 1)
LSP: (received) cpptools/initialize (id: 2)
LSP: (invoked) cpptools/initialize (id: 2)
cpptools version (TypeScript): 1.32.2
cpptools version (native): 1.32.2.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 2)
LSP: (received) cpptools/queryCompilerDefaults (id: 3)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 3)
Querying compiler for default C++ language standard using command line: /usr/bin/clang -x c++ -E -dM /dev/null
Detected language standard version: c++17
Querying compiler's default target using command line: "/usr/bin/clang" -dumpmachine
Compiler returned default target value: x86_64-pc-linux-gnu
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c17 -m64 -Wp,-v -fno-blocks -E -dM -x c /dev/null
Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/clang'
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c++17 -m64 -Wp,-v -fno-blocks -E -dM -x c++ /dev/null
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/clang'
LSP: Sending response (id: 3)
Processed c_cpp_properties.json in 0.00300528s
LSP: (received) cpptools/didChangeCppProperties (id: 4)
LSP: (invoked) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 5)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 6)
LSP: (received - deferred) cpptools/didChangeActiveEditor: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/resumeParsing
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 5)
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 6)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 7)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 8)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received) cpptools/preinitialize (id: 1)
LSP: (invoked) cpptools/preinitialize (id: 1)
LSP: Sending response (id: 1)
LSP: (received) cpptools/initialize (id: 2)
LSP: (invoked) cpptools/initialize (id: 2)
cpptools version (TypeScript): 1.32.2
cpptools version (native): 1.32.2.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 2)
LSP: (received) cpptools/queryCompilerDefaults (id: 3)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 3)
Querying compiler for default C++ language standard using command line: /usr/bin/clang -x c++ -E -dM /dev/null
Detected language standard version: c++17
Querying compiler's default target using command line: "/usr/bin/clang" -dumpmachine
Compiler returned default target value: x86_64-pc-linux-gnu
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c17 -m64 -Wp,-v -fno-blocks -E -dM -x c /dev/null
Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/clang'
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c++17 -m64 -Wp,-v -fno-blocks -E -dM -x c++ /dev/null
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/clang'
LSP: Sending response (id: 3)
Processed c_cpp_properties.json in 0.002697439s
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeCppProperties (id: 4)
LSP: (invoked) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 5)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 6)
LSP: (received - deferred) cpptools/didChangeActiveEditor: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/resumeParsing
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 5)
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 6)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 7)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 8)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received) cpptools/preinitialize (id: 1)
LSP: (invoked) cpptools/preinitialize (id: 1)
LSP: Sending response (id: 1)
LSP: (received) cpptools/initialize (id: 2)
LSP: (invoked) cpptools/initialize (id: 2)
cpptools version (TypeScript): 1.32.2
cpptools version (native): 1.32.2.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 2)
LSP: (received) cpptools/queryCompilerDefaults (id: 3)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 3)
Querying compiler for default C++ language standard using command line: /usr/bin/clang -x c++ -E -dM /dev/null
Detected language standard version: c++17
Querying compiler's default target using command line: "/usr/bin/clang" -dumpmachine
Compiler returned default target value: x86_64-pc-linux-gnu
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c17 -m64 -Wp,-v -fno-blocks -E -dM -x c /dev/null
Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/clang'
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c++17 -m64 -Wp,-v -fno-blocks -E -dM -x c++ /dev/null
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/clang'
LSP: Sending response (id: 3)
Processed c_cpp_properties.json in 0.00268936s
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeCppProperties (id: 4)
LSP: (invoked) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 5)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 6)
LSP: (received - deferred) cpptools/didChangeActiveEditor: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/resumeParsing
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 5)
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 6)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 7)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 8)
LSP: (received - deferred) cpptools/didChangeCompileCommands: file:///REDACTED_PATH/compile_commands.json
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received) cpptools/preinitialize (id: 1)
LSP: (invoked) cpptools/preinitialize (id: 1)
LSP: Sending response (id: 1)
LSP: (received) cpptools/initialize (id: 2)
LSP: (invoked) cpptools/initialize (id: 2)
cpptools version (TypeScript): 1.32.2
cpptools version (native): 1.32.2.0
Autocomplete is enabled.
Error squiggles are enabled.
Hover is enabled.
IntelliSense Engine = default.
LSP: Sending response (id: 2)
LSP: (received) cpptools/queryCompilerDefaults (id: 3)
LSP: (invoked) cpptools/queryCompilerDefaults (id: 3)
Querying compiler for default C++ language standard using command line: /usr/bin/clang -x c++ -E -dM /dev/null
Detected language standard version: c++17
Querying compiler's default target using command line: "/usr/bin/clang" -dumpmachine
Compiler returned default target value: x86_64-pc-linux-gnu
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c17 -m64 -Wp,-v -fno-blocks -E -dM -x c /dev/null
Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/clang'
Compiler info database not connected - skipping load.
Compiler query command line: /usr/bin/clang -std=c++17 -m64 -Wp,-v -fno-blocks -E -dM -x c++ /dev/null
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/clang'
LSP: Sending response (id: 3)
Processed c_cpp_properties.json in 0.002281911s
LSP: (received - deferred) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeCppProperties (id: 4)
LSP: (invoked) cpptools/didChangeCppProperties (id: 4)
LSP: (received - deferred) textDocument/didOpen: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 5)
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 6)
LSP: (received - deferred) cpptools/didChangeActiveEditor: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp
LSP: (received - deferred) cpptools/resumeParsing
LSP: (received - deferred) cpptools/getDocumentSymbols: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 7)
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 5)
LSP: $/cancelRequest (cpptools/getDocumentSymbols, id: 6)
LSP: (received - deferred) cpptools/getFoldingRanges: file:///REDACTED_PATH/REDACTED_PATH_TO_SOURCE.cpp (id: 8)
```

The `cpptools` log page (I have five of them once the two toast messages above show up) is quite sparse, only one line:

```
[Error - 21:04:22] The language server crashed. Restarting...
```

I'm running (once again...) an internal fork of Debian/Ubuntu.

**Comment by sean-mcmanus** [comment]
@sharadhr The cpptools process is crashing (not IntelliSense). It appears to crash while processing didChangeCppProperties. We attempt to log crash call stacks to a "C/C++ Crash Call Stacks" output pane if you see that. We did a potential fix for a crash in that function in May, but it hasn't been released yet, and without a call stack and I can't tell if it's the same root crash or not. This crash appears to happen earlier on, so it doesn't seem likely caused by a large compile_commands.json or an out of memory. One thing to check is to make sure no wal/shm next to the .db files exists, which is one case which could trigger this crash. If you set the `"browse":  { "databaseFilename": "${workspaceFolder}/.vscode/vc.db" } in c_cpp_properties.json then you more easily check the database files being used.

**Comment by optic5471ms** [comment]
I have had this problem in a 222MB file for a while, though it usually resolves itself after 30 seconds to a minute. However, in the last few days it has been a more than an hour long issue (if it resolves ever). The entire plugin just appears to freeze up. No intellisense, no swap to header/source, no go to symbol. The worst part is we use clang format and have format on save enabled. The plugin though is seemingly frozen at the initializing stage which causes the format to not happen and no file saving. At this point, the cpp plugin is basically useless in our project. We have 17K files, which turns into 4.5 million arguments across the whole project for the commands file. There was a recent update to vscode, though I dont know if that is to blame here. Reverting the plugin to the last release does not fix the issue. Making the commands file smaller does however resolve the issue, but it means that we dont have functionality across our codebase. I had copilot attempt to find a way around this and it made an extension to act as a provider which uses the commands file still. Im not sure that is the best solution, but it has for now unblocked me. I am wondering if this is something that can be fixed in the cpp plugin though.

**Comment by joelspadin-garmin** [comment]
I suspect I've been running into this same issue, as the projects I work on have 700+ MB compile_commands.json files, and everything provided by the C/C++ extension is incredibly slow. Code completion often takes 10+ seconds to show anything after opening a file. Alt+O to switch between source and header regularly takes so long that it's faster to just search for the file and open it manually. Go to definition is similarly slow, and doing a full text search on the codebase and sorting through the results is usually faster. I've been using https://github.com/xaverh/vscode-clang-format to work around the C/C++ extension locking up when trying to format on save, as it would often take 30+ seconds to save a file. I also occasionally get "the language server crashed" errors, but not so frequently that it gives up on restarting it.

After deleting the `compileCommands` entry from c_cpp_properties.json, then having my build system print out a list of compiler args, include paths, and defines and populating c_cpp_properties.json with that, things that would previously take 10-30 seconds now take 0-2 seconds. It could just be that it finished parsing everything, but adding and removing `compileCommands` seems to correlate pretty well with the performance issues.

**Comment by Colengms** [comment]
Hi @optic5471ms @joelspadin-garmin  

Can you confirm which version(s) of the C/C++ Extension you are using?  Are you seeing recent performance regressions with the latest C/C++ Extension preview/insiders builds (1.33.x)?

Would it be possible to provide a repro?  Depending on which compilers and compile options are involved, might point to different possible causes.

**Comment by joelspadin-garmin** [comment]
I have extension version 1.32.2. With `compileCommands` pointing to the large JSON file, pressing Alt+O to switch between header and source took 20-30 seconds on new file I opened. In one case, it took only 1-2 seconds to switch from header to source, but then it took ~30 seconds to switch back to the original header. Updating to the prerelease version 1.33.6 performed just as poorly, but it at least showed a progress notification for the 20-30 seconds instead of silently working in the background.

With `compileCommands` removed from c_cpp_properties.json, Alt+O took 0-2 seconds on every file I tested (picking a new file each time) on both the release and prerelease versions.

These performance issues have been going on for several months. I only recently noticed that compile_commands.json might be the cause.

This is in a proprietary codebase, so I can't provide a repro unfortunately. It should be okay for me to provide the compilers and options we use though.

We are currently using MSVC 14.44.35207 x64 with 
```
/nologo 
/MTd 
/showIncludes 
/FC
/Fd<differs per build target>.pdb 
/Z7 
/bigobj 
/Od 
/RTC1 
/TP 
/EHsc 
/std:c++17
<large number of /w flags which may differ for each file>
<~800 /I paths, send help>
<~40 /D defines>
```
and GCC versions 11.5.0 and newer, cross compiling for ARM 64 with
```
-mcpu=<varies by target hardware>
-march=<varies by target hardware>
-mbranch-protection=standard
--sysroot=<path to syroot from Yocto SDK>
-O2
-g
-pipe
-MMD
-fno-strict-aliasing
-fwrapv
-fstack-protector
-fstack-protector-strong
-fstack-clash-protection
-Og
-std=gnu++17
-pthread
<large number of -W flags which may differ for each file>
<~800 -I paths>
<~80 -D defines>
<~50 -isystem paths>
```

The MSVC compile_commands.json files are ~400 MB, and the GCC compile_commands.json files are ~700 MB. There are ~13000 entries in each file.

**Comment by Colengms** [comment]
Hi @joelspadin-garmin 

Thanks.  That helps a lot.

Are you using multiple `compile_commands.json` files in the same configuration, or are these two separate (MSVC vs GCC) config scenarios?

The C/C++ extension will invoke gcc/clang-based compilers to determine their system includes and predefined macros.  It should be stripping out arguments unrelated to that query (ones we know would not influence the system includes and predefines).  And, for any given set of arguments, the results are cached.  It's possible your command lines involve arguments that vary per-command-line, which we aren't stripping, leading to excessive compiler invocations.  You should be able to confirm by increasing logging (`"C_Cpp.loggingLevel": "Debug"`) and checking the output for an excessive number of compiler query invocations (and args used).  If we can identify args that vary between those invocations and determine them to be irrelevant to the query, we could add them to the list of args we filter out.

For MSVC/CL command lines, however, there is no need to invoke the compiler to determine system includes and predefines.  The overhead there should be primarily related to parsing the arguments (as some influence IntelliSense).  If you're seeing an issue involving only MSVC command lines, perhaps even some small some perf improvements there could be significant at scale.

**Comment by joelspadin-garmin** [comment]
These are two separate configurations, each with a single compile_commands.json. Both configs are compiling roughly the same set of files, with the GCC config building for Linux-based hardware and the MVSC config building a Windows version of the same application. As described in more detail later, the GCC configuration does include a small number of files that are compiled with MSVC instead of GCC, though I am seeing performance issues on both configs.

With the GCC config selected, I see several "compiler query command line" log messages that look like this:
```
Compiler query command line: C:\path\to\aarch64-poky-linux-g++.bat -mcpu=cortex-a76.cortex-a55 -march=armv8.2-a -mbranch-protection=standard -fstack-protector-strong -O2 --sysroot=<sysroot path> -g -pipe -fno-strict-aliasing -fwrapv -fstack-protector -fstack-clash-protection -Og -std=gnu++17 -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -pthread -Wp,-v -E -dM -x c++ nul
```
or the same thing with either `-std=gnu99` or `-std=c17` instead of `-std=gnu++17`. (I have no idea what our build system is doing that results in all those duplicate `-pthread` options.)

The GCC compiler won't run properly unless it has the correct environment variables, which are normally set up by the build system. (If run outside the correct environment, it will crash with no output.) To allow VS Code to query the compiler, I generated a .bat file which sets all the same environment variables as the build system and passes through the arguments to the GCC executable, then set that as the `compilerPath` in c_cpp_properties.json. That is why the compiler path is a .bat file rather than an .exe. However, the entries in compile_commands.json point to the GCC executable directly instead of this wrapper.

I did see one more thing that may be notable. The GCC/Linux config's compile_commands.json also contains some files that are compiled with CL.exe using MSVC options. There are a few source files in the build which compile Windows programs that are run to generate source files that are then compiled with GCC. I had not opened any of these files in a text editor in VS Code, but I saw messages for them appearing in the C/C++ extension's logs. These messages had a variety of different options such as

```
Compiler query command line: C:\path\to\aarch64-poky-linux-g++.bat /nologo /nologo /MTd /Fdobj.3.pdb /Z7 /bigobj /O2 /FC /c /FoC:\path\to\somefile.c.3.o -std=c17 -Wp,-v -E -dM -x c nul

Compiler query command line: C:\path\to\aarch64-poky-linux-g++.bat /nologo /nologo /MTd /Fdobj.4.pdb /Z7 /bigobj /Od /RTC1 /EHsc /std:c++17 /w14264 /w14628 /w14837 /w14905 /w14906 /w14928 /w14266 /w14355 /w14668 /w14706 /w14826 /w14242 /w14263 /w14302 /w14062 /w14254 /w14287 /w14289 /w14296 /w14311 /w14312 /w14339 /w14388 /w14412 /w14431 /w14437 /w14444 /w14471 /w14472 /w14536 /w14545 /w14546 /w14547 /w14549 /w14555 /w14557 /w14608 /w14682 /w14692 /w14931 /w14946 /w14962 /wd4018 /wd4057 /wd4065 /wd4100 /wd4121 /wd4127 /wd4152 /wd4204 /wd4242 /wd4244 /wd4245 /wd4267 /wd4296 /wd4305 /wd4310 /wd4324 /wd4355 /wd4388 /wd4389 /wd4800 /wd4668 /wd5287 /wd4054 /wd4055 /wd4206 /wd4250 /wd4351 /wd4396 /wd4512 /wd4201 /wd4510 /wd4610 /wd4701 /wd4703 /wd4995 /wd4996 /wd4505 /WX /FC /c /FoC:\path\to\somefile.cpp.4.o -std=gnu++17 -Wp,-v -E -dM -x c++ /dev/null
```

...which I assume is going to return garbage, but I don't know how relevant it is given that I'm not opening any of these files in the editor.

My c_cpp_properties.json configs look like this
```json
{
  "version": 4,
  "configurations": [
    {
      "name": "Windows",
      "intelliSenseMode": "windows-msvc-x64",
      "compilerPath": "c:\\program files\\microsoft visual studio\\2022\\professional\\vc\\tools\\msvc\\14.44.35207\\bin\\hostx64\\x64\\cl.exe",
      "compilerArgs": [ /* list of common args extracted from the build system */ ],
      "compileCommands": /* path to compile_commands.json */,
      "includePath": [ /* list of common include paths extracted from the build system */ ],
      "defines": [ /* list of common defines extracted from the build system */ ]
    },
    {
      "name": "Linux",
      "intelliSenseMode": "linux-gcc-arm64",
      "compilerPath": /* path to the .bat wrapper around GCC */,
      // compilerArgs, compileCommands, includePath, and defines set similarly to above
    }
  ]
}
```

**Comment by optic5471ms** [comment]
Hello. I am using extension version 1.33.6 in visual studio code 1.133. The issue does appear to only affect vscode 1.133, though the extension is slow in previous versions it is still usable. In the older vscode versions, it would take 2-30 seconds to Alt+O, save, find symbol, ... . In vscode 1.133 it seems to never resolve (left it alone for 2 hours without it passing the "Initializing..." stage). We are using clang to build with a vast number of definitions, includes, and compiler switches. The general setup for each file is as follows:
```json
{
  "directory": "The cmake build directory",
  "command": "path/to/clang/clang-cl.exe a hundred or so arguments",
  "file": "The file being compiled",
  "output": "The output obj file"
}
```

Some of the arguments we use:
```
Compiler arguments: --target=x86_64-pc-windows-msvc19.44 -c  -TP  -MD  -Z7  -WX  -w -std:c++14  -std:c++20 -ffp-contract=fast  -fno-pch-timestamp -Xclang -Zc:__cplusplus  -Zc:char8_t

Around another 50 warning and diagnostic control switches such as: -Wunused-parameter -Wunused-variable -Wno-clang-cl-pch -Wno-defaulted-function-deleted

Some msvc switches: /bigobj /EHsc /FC /fp:precise /GR /nologo /O2 /Ob2 /Oi /permissive- /sdl /W3 /W4 /w34996 /X /Zc:inline
And 2 dozen or so `/wd4005` style switches

Another 50 or so custom definitions.

A bunch of `-I` include path arguments
```

**Comment by Colengms** [comment]
Hi @optic5471ms 

Relatively little has changed in 1.33.x directly related to how `compile_commands.json` files are processed, so I suspect you might actually be encountering an issue with indirectly related changes.  Could you enable logging with `"C_Cpp.loggingLevel": "Debug"`, and monitor the `C/C++` output channel to see what it reports during that time?  It may be outputting some internal error diagnostics or otherwise showing slow progress on a particular type of task, such as gathering or parsing files.  If it stops abruptly, the last few things logged may point to something that is deadlocking, etc..  It would also be interesting to know whether (and which of) the 'cpptools.exe', 'cpptools-srv.exe', or 'cpptools-srv2.exe' processes remain busy during that hang, or effectively go idle, or if they're consuming an excessive amount memory.

**Comment by optic5471ms** [comment]
It appears from the logs that it is doing work, but it just is parsing every header and source file in all the Windows Kits on my program files folder and every file in the workspace. I imagine if I left it overnight it might actually finish. Until then the plugin is just non-functional it would seem. The `cpptools_srv.exe` is using 1.7GB of ram and climbing as it parses. `cpptools.exe` is up to nearly 900MB of ram though climbing slower than the server.
We were able to test that the vscode and cpptools versions dont seem to matter for this issue. Simply updating to the latest change in our repository causes the plugin to fail. So I would agree that it was not a recent change in either vscode or the cpptools plugin causing the issue.

**Comment by Colengms** [comment]
Hi @optic5471ms .  Most features of the C/C++ Extension should actually function while it's still ingesting files into the database.  (Some to a limited extent, such as Find All References).   Log output indicating that it's progressing through files actually seems like a healthy sign.  `cpptools-srv` is an IntelliSense process.  Its existence suggests that startup has completed and an IntelliSense translation unit has been successfully started for a file.  The `C/C++: Log Diagnostics` command can provide info on how that IntelliSense TU was configured, in case the issues you're seeing may be configuration related.  Despite the "Initializing..." language service status, could you single out some specific features/repros that are not working for you?
